### PR TITLE
Add target profile setting in VS Code

### PR DESCRIPTION
**Pull request by minestarks**
- A new configuration setting for target profile.
- Status bar item to show this setting.
- Command to set this setting.
- Language service update to consume this setting.

### CODE REVIEW

File: vscode/package.json [hunk 1]
```diff
@@ -125,7 +131,25 @@
           }
         }
       }
-    ]
+    ],
+    "configuration": {
+      "title": "Q#",
```
Comment (minestarks) [inline vscode/package.json:136]: Reviewer: If you're going to nitpick on the documentation here, please also nitpick on the documentation for the Python library as it's lifted directly from there.

https://github.com/microsoft/qsharp/blob/f2529e17d5ff21b189e846d456702ea077e989b8/pip/qsharp/_native.pyi#L7

File: vscode/src/common.ts [hunk 2]
```diff
@@ -1,18 +1,29 @@
...
   language: qsharpLanguageId,
   notebookType: "jupyter-notebook",
 };
+
+export function isQsharpDocument(document: TextDocument): boolean {
+  return languages.match(qsharpDocumentFilter, document) > 0;
+}
+
+export function isQsharpNotebookCell(document: TextDocument) {
+  return languages.match(qsharpNotebookCellDocumentFilter, document);
```
Comment (billti) [inline vscode/src/common.ts:24]: Why doesn't think return a bool if result if > 0 like the method above?

File: vscode/src/extension.ts [hunk 3]
```diff
@@ -124,21 +101,15 @@ function registerDocumentUpdateHandlers(languageService: ILanguageService) {
...
       }
     })
   );
 
   function updateIfQsharpDocument(document: vscode.TextDocument) {
-    if (
-      vscode.languages.match(qsharpDocumentFilter, document) &&
-      !vscode.languages.match(qsharpNotebookCellDocumentFilter, document)
-    ) {
+    if (isQsharpDocument(document) && !isQsharpNotebookCell(document)) {
```
Comment (billti) [inline vscode/src/extension.ts:112]: It is treated more like a bool here (per other comment)

File: vscode/src/extension.ts [hunk 4]
```diff
@@ -151,16 +122,81 @@ function registerDocumentUpdateHandlers(languageService: ILanguageService) {
...
+        .getConfiguration("qsharp")
+        .get<string>("targetProfile");
+      log.info(`Target profile changed: ${targetProfile}`);
+
+      switch (targetProfile) {
+        case "base":
+        case "full":
+          break;
+        default:
+          log.warn(`Invalid value for target profile: ${targetProfile}`);
```
Comment (billti) [inline vscode/src/extension.ts:189]: They may well have removed the setting, which is totally valid and shouldn't be a warning, so may want to not warn on 'undefined'

File: vscode/src/statusbar.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    refreshStatusBarItemValue();
+  }
+
+  function refreshStatusBarItemValue() {
+    // The target profile setting is a "window" scoped setting,
+    // meaning it can't be set on a per-folder basis. So we don't
+    // need to pass a specific scope here to retrieve the document
+    // value - we just use the workspace setting.
+    // VS Code will return the default value defined by the extension
+    // if none was set by the user, so targetProfile should always
```
Comment (billti) [inline vscode/src/statusbar.ts:57]: This is true if you have the configuration in the package.json and thus it contributes to the VS Code settings UX. I wonder if we need that UX though, being we have a command and status bar item to change it with anyway. Once Azure is on by default, we'd have no other settings in the UX (yet).

So maybe remove it from package.json for a cleaner UX? (And thus default value would be undefined).

File: vscode/package.json [hunk 6]
```diff
@@ -125,7 +131,25 @@
...
+      "properties": {
+        "qsharp.targetProfile": {
```
Comment (billti) [inline vscode/package.json:138]: I ended up using settings of the pattern "Q#.targetProfile", as the initial string shows up in the UX, and customers know the language as 'Q#', not 'Qsharp' (e.g. see below)

image

**Review by billti — APPROVED** [review]
Couple little things, else looks great. Thanks!

File: vscode/src/statusbar.ts (hunk 5 above)
Comment (minestarks) [reply vscode/src/statusbar.ts:57]: I'm not sure where we would persist it then. Can we just persist arbitrary configurations not defined in package.json? I wasn't sure of that.

File: vscode/package.json (hunk 6 above)
Comment (minestarks) [reply vscode/package.json:138]: Ohh, I was wondering where that string was coming from. Thanks for pointing this out.

File: vscode/src/statusbar.ts (hunk 5 above)
Comment (billti) [reply vscode/src/statusbar.ts:57]: Actually, I've been testing, and if you remove the 'configuration' contribution from package.json, while you can still read any arbitrary setting, you can't actually write one that's not defined!! So we need to leave this in package.json.

I renamed to "Q#.targetProfile" and integrated with my branch. If you check out my azure PR, this with changes is in there and works great!

**Comment by billti** [comment]
Just noticed a couple bugs in testing, the first to do with this PR, the second a more general one:

1) It doesn't set the profile on starting the language service, only when you change it. So if I have base profile set in my settings already, it doesn't take effect. I added this commit to my branch to fix it <https://github.com/microsoft/qsharp/pull/621/commits/0077b387b24720c93058112a24fe2c39a9eb9294>

2) In general, if you launch VS Code with a .qs file already loaded, no errors show up until you make an edit - be they profile errors, syntax errors, etc. Looks like we need to force a diagnostic pass if there are documents already open on launch. I can start looking into that too.

**Review by billti — CHANGES_REQUESTED** [review]
See latest feedback on issues. (Some already fixed in my branch).

**Comment by billti** [comment]
OK. The 2nd issue was pretty easy to fix (I think). Just needed to start listening for diagnostics earlier (before it adds the open diagnostics). Take a look at the below commit in my Azure branch and see if that looks right.

https://github.com/microsoft/qsharp/pull/621/commits/335d917c0e8a292dda9922bcfeccce0a679101fe

**Comment by billti** [comment]
This work was merged in as part of #621 . Thanks!

**Pull request closed without merging**